src/rez/solver.py

```diff
@@ -440,54 +850,90 @@ def get_intersection(self, range, max_packages=0):
-                package = value
-                if self.package_load_callback:
-                    self.package_load_callback(package)
+                if not isinstance(pkg, list):
+                    if self.package_load_callback:
+                        self.package_load_callback(pkg)
 
-                # access to timestamp causes a package load
-                if self.timestamp and package.timestamp > self.timestamp:
+                if self.timestamp and pkg.timestamp > self.timestamp:
                     continue
 
-                value = []
-                for var in package.iter_variants():
-                    requires = var.get_requires(build_requires=self.building)
+                # Remove the current package from the package_requests ..since it can not appear in its variants.
+                # probably no much gain and we are creating a new list.. might remove this line?
+                package_requests = [r for r in self.package_requests if r.name != pkg.name]
+
+                indexes.append(i)
+                variants_to_sort = []
+                original_variants = []
+                for var in pkg.iter_variants():
+                    requires = var.get_requires(
+                        build_requires=self.building)
+                    variants_to_sort.append(requires)
+                    original_variants.append(var)
+
+                if any(variants_to_sort):
+                    # timings.start("solver.sort_variants")
+                    sorted_variants = VariantSorter(variants_to_sort, package_requests).sort_variants()
+                    # timings.end("solver.sort_variants")
+                else:
+                    sorted_variants = variants_to_sort
+
+                for var in original_variants:
+                    # Map the variants to the new indexes in the sorted_variants
+                    original_requires = var.get_requires(build_requires=self.building)
                     userdata = var.handle.to_dict()
+                    new_index = var.index
+                    if new_index is not None:
+                        new_index = sorted_variants.index(original_requires)
+
                     variant = PackageVariant(name=self.package_name,
                                              version=var.version,
-                                             requires=requires,
-                                             index=var.index,
+                                             requires=original_requires,
+                                             index=new_index,
                                              userdata=userdata)
-                    value.append(variant)
-                entry[1] = value
-            variants.extend(value)
-            num_packages += 1
+                    loaded_variants.append(variant)
+                # sort all now by version and variant
+                loaded_variants = sorted(loaded_variants, key=lambda v: (v.version, v.index))
```

Comment:
I have differences in the code from line 911 to 923
our codes reads as it is on the resource2 branch  but with  `value = sorted(value, key=lambda v: (v.version, v.index))`  added before assigning it to  ` entry[1] = value`
```
                       value.append(variant)

                # sort all now by version and variant
                value = sorted(value, key=lambda v: (v.version, v.index))
                entry[1] = value
            variants.extend(value)
            num_packages += 1

```

with `value=[]` on line 895